### ISSUE TITLE
Don't require annotations of kwargs initialized to bool, float, int or string constants

### ISSUE THREAD

**Issue by edreamleo** [python/mypy#12352]
**Feature**

There should be no need for annotations for kwargs in the following (partial) function definitions:

```python
def bool_args(ok=True, die=False)
def int_arg(i=1)
def float_arg(ratio=1.0)
def string_arg(s="Hello World")
```

**Pitch**

At present, with disallow_untyped_defs = True and disallow_incomplete_defs = False, mypy complains that the signatures above lack annotations for the kinds of kwargs shown above.   Such annotations unnecessarily clutters code. It should be possible to infer the types of the kwargs based on the obvious types of the constants!

This feature would fix my biggest gripe about mypy :-)

**Comment by JelleZijlstra** [comment]
Pyright added this recently: microsoft/pyright#3092. I think adding inference here is reasonable, but a fully typed codebase should still add explicit annotations.

**Comment by edreamleo** [comment]
@JelleZijlstra Why? The annotations seem like pure cruft to me.

**Comment by JelleZijlstra** [comment]
There's no reason that the type of a default must be exactly the same as the accepted type. In the linked pyright thread, I provided an example.

**Comment by edreamleo** [comment]
@JelleZijlstra I never meant to imply that the type of the constant is *always* the correct annotation. This issue suggests only that it is reasonable, and often convenient, not to require an explicit type annotation for kwargs having specific kinds of initial values.

Yes, this default (implicit) annotation may lead to (valid) mypy complaints. For example, the "correct" annotation of a kwarg might be Optional[Bool], not Bool.  But in many cases the obvious default will suffice and mypy need not complain that a (pedantic) annotation has not been given.  Clear?

**Comment by KotlinIsland** [comment]
> There's no reason that the type of a default must be exactly the same as the accepted type

```py
a = 1 # is inferred
def f(a=1): # not inferred
  ...
```
Seems largely inconsistent to me.

**Comment by srittau** [comment]
I agree with OP that this would be a nice feature to have. (In fact, for TypeScript code I configure eslint to disallow redundant type annotations.) `def foo(x: int = 123)` seems like unnecessary churn and actually hard to read than `def foo(x=123)`. This doesn't mean that `def foo(x: int | str = 123)` isn't necessary sometimes.

**Comment by edreamleo** [comment]
Thanks for all the replies. I plan to start work on a PR in about a week.

**Comment by edreamleo** [comment]
I have begun work on PR #12918. I make no promises about completion date.

[This forum](https://github.com/edreamleo/ekr-mypy/discussions/1) contains my Engineering Notes for the project. I invite comments and suggestions there.

**Comment by edreamleo** [comment]
@JelleZijlstra I'm wondering whether you or some other guru would be willing to look at PR #12918 so see whether the changes to ASTConverter.make_argument have any chance of being correct?

Modulo some tracing code (which will go away) the new code is:
```python
        elif isinstance(default, ast3.Constant):  # #12352.
            arg_class = default.value.__class__.__name__
            if arg_class in ('bool', 'bytes', 'float', 'int', 'str'):
                arg_type = UnboundType(arg_class)
```
I have been dithering about whether to ask for help, but it seems like the best way forward here. Feedback would be much appreciated before doing a deep dive into unit tests. And yes, surely some tests *should* fail with the new code.

**Comment by edreamleo** [comment]
A heads up re schedule for this issue:

- I have been happy *using* mypy for several weeks, and expect that work to continue for at least a month.
- Several scheduling conflicts may push my work on this issue into next year.

P.S. A "love letter" to all you mypy devs: mypy is a superb tool and a superb feat of engineering :-) How did I ever live without it?

**Comment by KotlinIsland** [comment]
This is supported in basedmypy.

**Comment by edreamleo** [comment]
@KotlinIsland Thanks for the heads up. Do you plan to issue a PR for mypy itself?

**Comment by KotlinIsland** [comment]
@edreamleo Not at this time, I have much more pressing items. But I'd be happy for anyone to port anything from basedmypy into mypy.

**Comment by edreamleo** [comment]
@KotlinIsland Alright. I'll have a look at your commits :-)

**Comment by edreamleo** [comment]
@JelleZijlstra It's been over a year since I opened this issue. I'm not sure I'll get to it in any timely fashion. Shall we just close it?

**Comment by KotlinIsland** [comment]
@edreamleo I don't see why the issue should be closed just because noone is working on it.

**Comment by edreamleo** [comment]
@KotlinIsland Thanks for your comment. I'll leave it open and won't feel guilty about not completing it :-)
